src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -665,3 +674,120 @@ edgeSet = Set.fromAscList . edgeList
 -- @
 consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
 consistent (BAM lr rl) = internalEdgeList lr == sort (map Data.Tuple.swap $ internalEdgeList rl)
+
+
+data PartId = LeftPart | RightPart
+    deriving (Eq, Show)
+
+
+otherPart :: PartId -> PartId
+otherPart LeftPart  = RightPart
+otherPart RightPart = LeftPart
+
+
+-- | Test bipartiteness of given graph. In case of success, return an
+-- `AdjacencyMap` with the same set of edges and the vertices marked with the
+-- part they belong to.
+--
+-- /Note/: as `AdjacencyMap` only represents __undirected__ bipartite graphs,
+-- all edges in the input graph are assumed to be bidirected and all edges in
+-- the output `AdjacencyMap` are bidirected.
+--
+-- It is advised to use 'leftVertexList' and 'rightVertexList' to obtain the
+-- partition of the vertices and 'hasLeftVertex' and 'hasRightVertex' to check
+-- whether a vertex belong to a part.
+--
+-- Complexity: /O((n + m) log(n))/.
+--
+-- @
+-- detectParts 'Algebra.Graph.AdjacencyMap.empty'                             == Just 'empty'
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.vertex' x))               == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.edge' x y))               == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.edges' [(1, 2), (1, 3)])) == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.clique' x))               == x < 3
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.star' x ys))              == not (x `elem` ys)
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.biclique' xs ys))         == True
+-- 'Data.Maybe.isJust' ('fromBipartite' ('toBipartite' am))       == True
+-- 'Data.Maybe.isJust' (detectParts (1 * 2 * 3))              == False
+-- 'Data.Maybe.isJust' (detectParts ((1 + 2) * (3 + 4)))      == True
+-- @
+detectParts :: Ord a => AM.AdjacencyMap a -> Maybe (AdjacencyMap a a)
```

Comment:
How does this function behave if the answer is not unique? Is the result optimal in some sense, e.g. lexicographically smallest?

Reply:
*Any* result for a non-empty graph is not unique, as we can always swap the parts.

I didn't think of a completely formal proof, but it seems to me that the bit string of the part identifiers for each vertex is lexicographically minimal. It seems that for a connected non-empty graph there are always exactly two unique answers. If so, then it is correct: each component is visited in its minimal vertex for the first time and this vertex is put into the left part.

Reply:
Sounds good, thanks. Let's document this in comments, and perhaps, add an example and test that demonstrate this.

Reply:
Perhaps, the existing examples are already sufficient, in which case you can just say "see examples" in the comment :)

Reply:
They are not, as for now we know it, we can add something like
```
detectParts (1 * 2 + 3 * 4) = Just (edges [(1, 2), (3, 4)])
```
This seems to look better than
```
isJust (detectParts (1 * 2 + 3 * 4)) == True
``

Reply:
Indeed, `detectParts (1 * 2 + 3 * 4) = Just (edges [(1, 2), (3, 4)])` is much more informative.